doc: mention termination_by and decreasing_by (#3016)
leanprover/lean4: doc/declarations.md

| _   => 3
```

The case where patterns are matched against an argument whose type is an inductive family is known as *dependent pattern matching*. This is more complicated, because the type of the function being defined can impose constraints on the patterns that are matched. In this case, the equation compiler will detect inconsistent cases and rule them out.

```lean
universe u

inductive Vector (α : Type u) : Nat → Type u
| nil  : Vector α 0
| cons : α → Vector α n → Vector α (n+1)

namespace Vector

def head {α : Type} : Vector α (n+1) → α
| cons h t => h

def tail {α : Type} : Vector α (n+1) → Vector α n
| cons h t => t

def map {α β γ : Type} (f : α → β → γ) :
  ∀ {n}, Vector α n → Vector β n → Vector γ n
| 0,   nil,       nil       => nil
| n+1, cons a va, cons b vb => cons (f a b) (map f va vb)

> end Vector

Author: Yuck, `autoImplicits` :-)

Is that something we want to encourage, especially in places where people do _not_ have the benefit of IDE support?

> end Vector

Reviewer: I don't think there is any indication that lean discourages autoimplicits. That's a mathlib thing. (Besides this, I think for type arguments they are quite unobjectionable, being the default behavior in haskell and ocaml.)

> end Vector

Reviewer 2: I agree - and especially in a context like this, decreasing syntactic noise seems useful.

```

.. _recursive_functions:

Recursive functions
===================

Lean has to ensure that a recursive function definition is terminating, and has two strategies for that.

Structural recursion
--------------------

> If the definition of a function contains recursive calls, the equation compiler tries to interpret the definition as a structural recursion. In order for that to succeed, the recursive arguments must be subterms of the corresponding arguments on the left-hand side. The function is then defined using a *course of values* recursion, using automatically generated functions ``below`` and ``brec`` in the namespace corresponding to the inductive type of the recursive argument. In this case the defining equations hold definitionally, possibly with additional case splits.

Author: Previouisly, all of this was documented under “equation compiler”, and so far I don’t have good name for the component that processes recursive definitions (besides “derecursifier”, but I doubt that we should really use this.). It may make sense to avoid giving this component a name alltogether.

Feel free to suggest better wordings! But remember that this is a MVP-doc-PR to have something, and maybe more serious copyediting should happen as part of a more complete overhaul of the documentation, and maybe by a professional technical writer.

> If the definition of a function contains recursive calls, the equation compiler tries to interpret the definition as a structural recursion. In order for that to succeed, the recursive arguments must be subterms of the corresponding arguments on the left-hand side. The function is then defined using a *course of values* recursion, using automatically generated functions ``below`` and ``brec`` in the namespace corresponding to the inductive type of the recursive argument. In this case the defining equations hold definitionally, possibly with additional case splits.

Reviewer 2: I like not mentioning a component name. What about "Lean" instead of "the equation compiler" here? I think that's just as clear, and less likely to throw people off.

> If the definition of a function contains recursive calls, the equation compiler tries to interpret the definition as a structural recursion. In order for that to succeed, the recursive arguments must be subterms of the corresponding arguments on the left-hand side. The function is then defined using a *course of values* recursion, using automatically generated functions ``below`` and ``brec`` in the namespace corresponding to the inductive type of the recursive argument. In this case the defining equations hold definitionally, possibly with additional case splits.

Author: ```suggestion
If the definition of a function contains recursive calls, Lean tries to interpret the definition as a structural recursion. In order for that to succeed, the recursive arguments must be subterms of the corresponding arguments on the left-hand side.

The function is then defined using a *course of values* recursion, using automatically generated functions ``below`` and ``brec`` in the namespace corresponding to the inductive type of the recursive argument. In this case the defining equations hold definitionally, possibly with additional case splits.
```

| _   => 3
```

The case where patterns are matched against an argument whose type is an inductive family is known as *dependent pattern matching*. This is more complicated, because the type of the function being defined can impose constraints on the patterns that are matched. In this case, the equation compiler will detect inconsistent cases and rule them out.

```lean
universe u

inductive Vector (α : Type u) : Nat → Type u
  | nil  : Vector α 0
  | cons : α → Vector α n → Vector α (n+1)

namespace Vector

def head : Vector α (n+1) → α
  | cons h t => h

def tail : Vector α (n+1) → Vector α n
  | cons h t => t

def map (f : α → β → γ) : Vector α n → Vector β n → Vector γ n
  | nil, nil => nil
  | cons a va, cons b vb => cons (f a b) (map f va vb)

end Vector
```

.. _recursive_functions:

Recursive functions
===================

> Lean must ensure that a recursive functions terminate, for which there are two strategies: _structural recursion_, in which all recursive calls are made on smaller parts of the input data, and _well-founded recursion_, in which recursive calls are justified by showing that arguments to recursive calls are smaller according to some other measure.

Reviewer 3: ```suggestion
Lean must ensure that a recursive function terminates, for which there are two strategies: _structural recursion_, in which all recursive calls are made on smaller parts of the input data, and _well-founded recursion_, in which recursive calls are justified by showing that arguments to recursive calls are smaller according to some other measure.
```

Structural recursion
--------------------

> If the definition of a function contains recursive calls, Lean tries to interpret the definition as a structural recursion. In order for that to succeed, the recursive arguments must be subterms of the corresponding arguments on the left-hand side.

Reviewer 3: ```suggestion
If the definition of a function contains recursive calls, Lean first tries to interpret the definition as a structural recursion. In order for that to succeed, the recursive arguments must be subterms of the corresponding arguments on the left-hand side.
```
Helpfully foreshadows that there is a fall-back?

end Hide
```

> Well-founed recursion

Reviewer 3: ```suggestion
Well-founded recursion
```

---------------------

> If structural recursion fails, the equation compiler falls back on well-founded recursion. It tries to infer an instance of ``SizeOf`` for the type of each argument, and then tries to find a permutation of the arguments such that each recursive call is decreasing under the lexicographic order with respect to ``sizeOf`` measures.  Lean uses information in the local context, so you can often provide the relevant proof manually using ``have`` in the body of the definition. In this case of well-founded recursion, the defining equations hold only propositionally, and can be accessed using ``simp`` and ``rewrite`` with the name ``foo``.

Author: ```suggestion
If structural recursion fails, the equation compiler falls back on well-founded recursion. It tries to infer an instance of ``SizeOf`` for the type of each argument, and then tries to find a permutation of the arguments such that each recursive call is decreasing under the lexicographic order with respect to ``sizeOf`` measures.  Lean uses information in the local context, so you can often provide the relevant proof manually using ``have`` in the body of the definition.

In the case of well-founded recursion, the equation used to declare the function holds only propositionally, but not definitionally, and can be accessed using ``unfold``, ``simp`` and ``rewrite`` with the function name (for example ``unfold foo`` or ``simp [foo]``, where ``foo`` is the function defined with well-founded recursion).
```

(Not perfect, just aiming for good enough at this point…)

end Hide
```

> If Lean cannot find a permutation of the arguments for which all recursive calls are decreasing, it will print a table that contains, for every recursive call, which arguments Lean could prove to be decreasing.

Reviewer 3: Would it make sense to include an example table here? Or too difficult to keep in sync?

Author: Will do

> Lean will print the termination argument it found if ``set_option showInferredTerminationBy true`` is set.

Author: For the table I wrote it while you posted.

Adding an example for the output from ``showInferredTerminationBy`` here is almost not helpful, because the syntax it is presented in is discussed later, and clearly recognizable as that.

> Lean will print the termination argument it found if ``set_option showInferredTerminationBy true`` is set.

Reviewer 2: Sounds good

> If Lean does not find the termination argument, or if you want to be explict, you can append a `termination_by` clause to the function definition. It is fo the form

Reviewer 3: ```suggestion
If Lean does not find the termination argument, or if you want to be explict, you can append a `termination_by` clause to the function definition. It is of the form
```

```
termination_by f a₁ … aₙ => e
```

> where ``f`` is the function name (or ``_``), ``a₁ … aₙ``  are the parameters (or ``_``), and ``e`` is the expression that should be decreasing at each recursive call. The the type of `e` should be an instance of the class ``WellFoundedRelation``, which determins how to compare two values of that type.

Reviewer 3: ```suggestion
where ``f`` is the function name (or ``_``), ``a₁ … aₙ``  are the parameters (or ``_``), and ``e`` is the expression that should be decreasing at each recursive call. The the type of `e` should be an instance of the class ``WellFoundedRelation``, which determines how to compare two values of that type.
```

> where ``f`` is the function name (or ``_``), ``a₁ … aₙ``  are the parameters (or ``_``), and ``e`` is the expression that should be decreasing at each recursive call. The the type of `e` should be an instance of the class ``WellFoundedRelation``, which determins how to compare two values of that type.

Reviewer 3: ```suggestion
where ``f`` is the function name (or ``_``), ``a₁ … aₙ``  are the parameters (or ``_``), and ``e`` is the expression that should be decreasing at each recursive call. The type of `e` should be an instance of the class ``WellFoundedRelation``, which determins how to compare two values of that type.
```

> where ``f`` is the function name (or ``_``), ``a₁ … aₙ``  are the parameters (or ``_``), and ``e`` is the expression that should be decreasing at each recursive call. The the type of `e` should be an instance of the class ``WellFoundedRelation``, which determins how to compare two values of that type.

Author: ```suggestion
where ``f`` is the function name (or ``_``), ``a₁ … aₙ``  are the parameters (or ``_``), and ``e`` is the expression that should be decreasing at each recursive call. The type of `e` should be an instance of the class ``WellFoundedRelation``, which determins how to compare two values of that type.
```

By default, Lean uses the tactic ``decreasing_tactic`` when proving that an argument is decreasing; see its documentation for how to globally extend it. You can also choose to use a different tactic for a given function definition with the clause
```
decreasing_by <tac>
```

> which should come ``termination_by`, if present.

Reviewer 3: ```suggestion
which should come after ``termination_by`, if present.
```

end Hide
```

> If Lean cannot find a permutation of the arguments for which all recursive calls are decreasing, it will print a table that contains, for every recursive call, which arguments Lean could prove to be decreasing. For example, a function with three recursive calls and four paramters might cause the following message to be printed

Reviewer 2: ```suggestion
If Lean cannot find a permutation of the arguments for which all recursive calls are decreasing, it will print a table that contains, for every recursive call, which arguments Lean could prove to be decreasing. For example, a function with three recursive calls and four parameters might cause the following message to be printed
```

```
example.lean:37:0-43:31: error: Could not find a decreasing measure.
The arguments relate at each recursive call as follows:
(<, ≤, =: relation proved, ? all proofs failed, _: no proof attempted)
           x1 x2 x3 x4
1) 39:6-27  =  =  _  =
2) 40:6-25  =  ?  _  <
3) 41:6-25  <  _  _  _
Please use `termination_by` to specify a decreasing measure.
```

This table should be read as follows:

 * The first recursive call, in line 39, arguments 1, 2 and 4 are equal to the function's paramters.

>  * The second recursive call, in line 40, has an equal first, a smaller forth argument, and nothing could be inferred for the second argument.

Reviewer 2: ```suggestion
 * The second recursive call, in line 40, has an equal first, a smaller fourth argument, and nothing could be inferred for the second argument.
```

 * The third recursive call, in line 41, has a decreasing first argument.

>  * No other proofs were attempted, either because the paramter has a type without a non-trivial ``WellFounded`` instance (parameter 3), or because it is already clear that no decreasing measure can be found.

Reviewer 2: ```suggestion
 * No other proofs were attempted, either because the parameter has a type without a non-trivial ``WellFounded`` instance (parameter 3), or because it is already clear that no decreasing measure can be found.
```

end Hide
```

If Lean cannot find a permutation of the arguments for which all recursive calls are decreasing, it will print a table that contains, for every recursive call, which arguments Lean could prove to be decreasing. For example, a function with three recursive calls and four parameters might cause the following message to be printed

```
example.lean:37:0-43:31: error: Could not find a decreasing measure.
The arguments relate at each recursive call as follows:

> (<, ≤, =: relation proved, ? all proofs failed, _: no proof attempted)

Reviewer 3: Colons are inconsistent here: should there be one after the `?`?

The colons are't that great, in any case, as I'm tempted to misinterpret `=:` as an atom.

Author: Good point. How about putting the prose first and using parentheses?
```suggestion
Relation proved (<, ≤, =), all proofs failed (?), no proof attempted (_)
```

           x1 x2 x3 x4
1) 39:6-27  =  =  _  =
2) 40:6-25  =  ?  _  <
3) 41:6-25  <  _  _  _
Please use `termination_by` to specify a decreasing measure.
```

This table should be read as follows:

>  * The first recursive call, in line 39, arguments 1, 2 and 4 are equal to the function's paramters.

Reviewer 3: ```suggestion
 * The first recursive call, in line 39, arguments 1, 2 and 4 are equal to the function's parameters.
```

>  * The first recursive call, in line 39, arguments 1, 2 and 4 are equal to the function's paramters.

Author: ```suggestion
 * In the first recursive call, in line 39, arguments 1, 2 and 4 are equal to the function's parameters.
```

>  * The second recursive call, in line 40, has an equal first, a smaller fourth argument, and nothing could be inferred for the second argument.

Reviewer 3: ```suggestion
 * The second recursive call, in line 40, has an equal first argument, a smaller fourth argument, and nothing could be inferred for the second argument.
```

 * The third recursive call, in line 41, has a decreasing first argument.
 * No other proofs were attempted, either because the parameter has a type without a non-trivial ``WellFounded`` instance (parameter 3), or because it is already clear that no decreasing measure can be found.

Lean will print the termination argument it found if ``set_option showInferredTerminationBy true`` is set.

> If Lean does not find the termination argument, or if you want to be explict, you can append a `termination_by` clause to the function definition, after the function's body, but before the `where` clause if present. It is of the form

Reviewer 3: ```suggestion
If Lean does not find the termination argument, or if you want to be explicit, you can append a `termination_by` clause to the function definition, after the function's body, but before the `where` clause if present. It is of the form
```

```
termination_by e
```
where ``e`` is an expression that depends on the parameters of the function and should be decreasing at each recursive call. The type of `e` should be an instance of the class ``WellFoundedRelation``, which determines how to compare two values of that type.

> If ``f`` has parameters “after the ``:``” (a `fun` after the `:=`, or function equations), these must be given names using the syntax
> ```
> termination_by a₁ … aₙ => e
> ```

Author: ````suggestion
If ``f`` has parameters “after the ``:``” (for example when defining functions via patterns using `|`), then these can be brought into scope using the syntax.
```
termination_by a₁ … aₙ => e
```
````

No longer a ”must”.

end
```

> In a set of mutually recursive function, either all or no functions must have an explicit termination argument (``termination_by``). A change of the default termination tactic (``decreasing_by``) only affects the proofs about the recursive calls of that function, not the other functiosn in the group.

Reviewer 2: ```suggestion
In a set of mutually recursive function, either all or no functions must have an explicit termination argument (``termination_by``). A change of the default termination tactic (``decreasing_by``) only affects the proofs about the recursive calls of that function, not the other functions in the group.
```

```
mutual
theorem even_of_odd_succ : ∀ n, Odd (n + 1) → Even n
| _, odd_succ n h => h
termination_by n h => h
decreasing_by decreasing_tactic

theorem odd_of_even_succ : ∀ n, Even (n + 1) → Odd n
| _, even_succ n h => h
termination_by n h => h
end
```

Another way to express mutual recursion using local function definitions in ``where`` or ``let rec`` clauses: these can be mutually recursive with each other and their containing function:

```

> theorem even_of_odd_succ : ∀ n, Odd (n + 1) → Even n
> | _, odd_succ n h => h
> termination_by n h => h
>   where
>     theorem odd_of_even_succ : ∀ n, Even (n + 1) → Odd n
>     | _, even_succ n h => h
>     termination_by n h => h

Reviewer 2: We typically indent the cases:
```suggestion
theorem even_of_odd_succ : ∀ n, Odd (n + 1) → Even n
  | _, odd_succ n h => h
termination_by n h => h
  where
    theorem odd_of_even_succ : ∀ n, Even (n + 1) → Odd n
      | _, even_succ n h => h
    termination_by n h => h
```

end
```

In a set of mutually recursive function, either all or no functions must have an explicit termination argument (``termination_by``). A change of the default termination tactic (``decreasing_by``) only affects the proofs about the recursive calls of that function, not the other functions in the group.

```
mutual
theorem even_of_odd_succ : ∀ n, Odd (n + 1) → Even n
| _, odd_succ n h => h
termination_by n h => h
decreasing_by decreasing_tactic

theorem odd_of_even_succ : ∀ n, Even (n + 1) → Odd n
| _, even_succ n h => h
termination_by n h => h
end
```

> Another way to express mutual recursion using local function definitions in ``where`` or ``let rec`` clauses: these can be mutually recursive with each other and their containing function:

Reviewer 3: ```suggestion
Another way to express mutual recursion is using local function definitions in ``where`` or ``let rec`` clauses: these can be mutually recursive with each other and their containing function:
```
